### ISSUE TITLE
Mismatched type errors have become useless

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#39343]
A minimal test case. My actual case is much more complex, involves functions/callbacks with multiple params and several layers of typedefs.

```rust
fn takes_cb(f: fn(i8)) {}

fn main() {
    fn callback(x: i32) {}
    takes_cb(callback)
}
```

The error reported on 1.11.0 is this:

```
test.rs:5:14: 5:22 error: mismatched types [E0308]
test.rs:5     takes_cb(callback)
                       ^~~~~~~~
test.rs:5:14: 5:22 help: run `rustc --explain E0308` to see a detailed explanation
test.rs:5:14: 5:22 note: expected type `fn(i8)`
test.rs:5:14: 5:22 note:    found type `fn(i32) {main::callback}`
```

Useful and obvious. What error looks like now (stable (1.14.0)/beta/nightly (1.16.0))?

```
error[E0308]: mismatched types
 --> test.rs:5:14
  |
5 |     takes_cb(callback)
  |              ^^^^^^^^ expected i8, found i32

```

Cute but is as worthless and confusing as it gets.

**Comment by eddyb** [comment]
This actually looks like a change in coercion? That is, a mismatch in the signature used to prevent the coercion, resulting in the subtyping fallback which simply fails, at the othermost level, whereas now the coercion is attempted and the error *from inside the signature mismatch* is being reported.

**Comment by nikomatsakis** [comment]
@eddyb any clues what change might have caused that?

cc @estebank @jonathandturner -- this doesn't look familiar, does it?

**Comment by jonathandturner** [comment]
@nikomatsakis - not to me. I thought we still always showed the expected/found full type since sometimes you need it to understand the error (this one included)

**Comment by estebank** [comment]
@nikomatsakis sadly, it doesn't for me either.

@jonathandturner [it doesn't if the type mismatch is between primitive types](https://github.com/rust-lang/rust/blob/master/src/librustc/infer/error_reporting.rs#L579-L584), which leads me to think that somewhere a `TypeError` is being constructed for the arguments alone, and not for the method signature. Probably a callee of [`report_mismatched_types`](https://github.com/rust-lang/rust/blob/master/src/librustc/infer/mod.rs#L1458-L1466).

**Comment by arielb1** [comment]
triage: P-high - serious diagnostic regression, assign to @pnkfelix.

**Comment by estebank** [comment]
While doing something else around type errors, I realized by chance why this error was happening.

**Comment by pnkfelix** [comment]
(seems like #39905 will fix this.)
